ambry-api/src/main/java/com.github.ambry/rest/RestUtils.java

```diff
@@ -589,6 +589,39 @@ public static String getOperationOrBlobIdFromUri(RestRequest restRequest, RestUt
     return subResource;
   }
 
+  /**
+   * When validating secure path is required by container, this method parses the secure path from uri.
+   * Getting secure path is based on the assumption that the first non-empty segment (split by "/") is secure path.
+   * @param restRequest {@link RestRequest} containing metadata about the request.
+   * @return secure path if uri contains one; otherwise empty string is returned.
+   */
+  public static String getSecurePath(RestRequest restRequest) {
+    String path = restRequest.getPath();
+    SubResource subResource = getBlobSubResource(restRequest);
+    int endIndex = path.length();
+    String securePath = "";
+    if (subResource != null) {
+      // remove sub resource from path
+      endIndex = endIndex - subResource.name().length() - 1;
+      path = path.substring(0, endIndex);
+    }
+    if (path.startsWith("/")) {
+      // remove leading slash
+      path = path.substring(1);
+    }
+    // if this is a signed id, return empty secure path
+    if (!path.startsWith(RestUtils.SIGNED_ID_PREFIX)) {
```

Comment:
As we discussed offline, there is no reason that a signed ID cannot have a secure path prefix ahead of it.

For example, I think that calling this method on `/secure-path/signedId/ABCDEF` should return `secure-path`. 

I think that we can actually combine most of the logic of `getOperationOrBlobId` and `getSecurePath`. Let's see if we can make a helper method that returns a `Pair<String,String>` that includes the path prefix and operationOrBlobId. This new method could be called directly in `SecurityService` since you need both the path prefix and operationOrBlobId and we can remove the `getSecurePath` method.

```java
  public static Pair<String, String> getPrefixAndResourceFromUri(RestRequest restRequest, SubResource subResource,
      List<String> prefixesToRemove) {
    String path = restRequest.getPath();
    int startIndex = 0;
    // remove query string.
    int endIndex = path.indexOf("?");
    if (endIndex == -1) {
      endIndex = path.length();
    }
    // remove prefix.
    String prefixFound = "";
    if (prefixesToRemove != null) {
      for (String prefix : prefixesToRemove) {
        if (path.startsWith(prefix)) {
          prefixFound = prefix;
          startIndex = prefix.length();
          break;
        }
      }
    }
    // remove subresource if present.
    if (subResource != null) {
      // "- 1" removes the "slash" that precedes the sub-resource.
      endIndex = endIndex - subResource.name().length() - 1;
    }
    String operationOrBlobId = path.substring(startIndex, endIndex);
    if ((operationOrBlobId.isEmpty() || operationOrBlobId.equals("/")) && restRequest.getArgs()
        .containsKey(Headers.BLOB_ID)) {
      operationOrBlobId = restRequest.getArgs().get(Headers.BLOB_ID).toString();
    }
    return new Pair<>(prefixFound, operationOrBlobId);
  }

  public static String getOperationOrBlobIdFromUri(RestRequest restRequest, SubResource subResource,
      List<String> prefixesToRemove) {
    return getPrefixAndResourceFromUri(restRequest, subResource, prefixesToRemove).getSecond();
  }
```